model/classes/ResultService.php

```diff
@@ -0,0 +1,131 @@
+                        <imsx_messageRefIdentifier>{{messageIdentifier}}</imsx_messageRefIdentifier>
+                        <imsx_operationRefIdentifier>replaceResult</imsx_operationRefIdentifier>
+                    </imsx_statusInfo>
+                </imsx_POXResponseHeaderInfo>
+            </imsx_POXHeader>
+            <imsx_POXBody>
+                <replaceResultResponse />
+            </imsx_POXBody>
+        </imsx_POXEnvelopeResponse>
+    ';
+    const SCORE_DESCRIPTION_TEMPLATE = 'Score for {{sourceId}} is now {{score}}';
+
+    private $dom;
+    public static $statuses = array(
+        400 => 'Invalid score',
+        404 => 'DeliveryExecution not found',
+        501 => 'Method not implemented (for other requests than replaceResultRequest)',
+        500 => 'Other internal error => client should retry',
+    );
+
+    /**
+     * @param $payload string
+     * @return array|null
+     */
+    public function loadPayload($payload)
+    {
+        try {
+            $this->dom = new \DOMDocument();
+            $this->dom->loadXML($payload);
+            $this->validateResultRequest($payload);
+        } catch (\Exception $e) {
+//            $this->logError('Request XML does not have replaceResultRequest element');
+            return [
+                '{{codeMajor}}' => self::FAILURE_MESSAGE,
+                '{{description}}' => self::$statuses[501],
+                '{{messageId}}' => '',
+            ];
+        }
+    }
+
+    public function getResult()
+    {
+        $xpath = new \DOMXPath($this->dom);
+        $xpath->registerNamespace('lti', "http://www.imsglobal.org/services/ltiv1p1/xsd/imsoms_v1p0");
+
+        $messageIdentifier = $xpath->evaluate('/lti:imsx_POXEnvelopeRequest/lti:imsx_POXHeader/lti:imsx_POXRequestHeaderInfo/lti:imsx_messageIdentifier')
+            ->item(0)->nodeValue;
+        $score = $xpath->evaluate('/lti:imsx_POXEnvelopeRequest/lti:imsx_POXBody/lti:replaceResultRequest/lti:resultRecord/lti:result/lti:resultScore/lti:textString')
+            ->item(0)->nodeValue;
+        $sourcedId = $xpath->evaluate('/lti:imsx_POXEnvelopeRequest/lti:imsx_POXBody/lti:replaceResultRequest/lti:resultRecord/lti:sourcedGUID/lti:sourcedId')
+            ->item(0)->nodeValue;
+
+        return [
+            'messageIdentifier' => $messageIdentifier,
+            'score' => $score,
+            'sourcedId' => $sourcedId,
+        ];
+    }
+
+    public function isScoreValid($score)
+    {
+        return (is_numeric($score) && $score >= 0 && $score <= 1);
+    }
+
+    private function validateResultRequest()
+    {
+        $xpath = new \DOMXPath($this->dom);
+        $xpath->registerNamespace('lti', "http://www.imsglobal.org/services/ltiv1p1/xsd/imsoms_v1p0");
+        $elements = $xpath->evaluate('/lti:imsx_POXEnvelopeRequest/lti:imsx_POXBody/lti:replaceResultRequest');
+
+        if ($elements->length === 0) {
+            throw new \Exception('Request XML does not have replaceResultRequest element');
```

Comment:
This message is never uesd since you catch the exception and write another in loadPayload.

Reply:
@julien-sebire  It goes into it if the XML is correct but lti:replaceResultRequest missed